### PR TITLE
13.0 stock_partner_delivery_window: Handle working days

### PR DESCRIPTION
**Pull request by mmequignon**
Adds a new choice for partner's delivery_time_preference to raise an error when the picking scheduled date is set to a weekend day.

### CODE REVIEW

File: stock_partner_delivery_window/models/res_partner.py [hunk 1]
```diff
@@ -49,19 +57,13 @@ def get_delivery_windows(self, day_name=None):
         :param day: The day name (see time.weekday, ex: 0,1,2,...)
         :return: dict partner_id: delivery_window recordset
         """
-        res = {}
+        self.ensure_one()
         domain = [("partner_id", "in", self.ids)]
```
Comment (i-vyshnevska) [inline stock_partner_delivery_window/models/res_partner.py]:
```suggestion
        domain = [("partner_id", "=", self.id)]
```

File: stock_partner_delivery_window/models/stock_picking.py [hunk 2]
```diff
@@ -13,31 +13,39 @@ def _planned_delivery_date(self):
...
+        formatted_scheduled_date = format_datetime(self.env, scheduled_date)
+        partner = self.partner_id
+        if partner.delivery_time_preference == "workdays":
+            message = _(
+                "The scheduled date is {} ({}), but the partner is "
+                "set to prefer deliveries on working days."
+            ).format(formatted_scheduled_date, scheduled_date.weekday())
+        else:
+            delivery_windows_strings = []
+            for w in self.partner_id.get_delivery_windows():
```
Comment (i-vyshnevska) [inline stock_partner_delivery_window/models/stock_picking.py]:
```suggestion
            for w in partner_id.get_delivery_windows():
```

File: stock_partner_delivery_window/models/stock_picking.py [hunk 3]
```diff
@@ -13,31 +13,39 @@ def _planned_delivery_date(self):
...
+                delivery_windows_strings.append(
+                    "  * {} ({})".format(w.display_name, self.partner_id.tz)
```
Comment (i-vyshnevska) [inline stock_partner_delivery_window/models/stock_picking.py]:
```suggestion
                    "  * {} ({})".format(w.display_name, partner_id.tz)
```

File: stock_partner_delivery_window/README.rst [hunk 4]
```diff
@@ -44,6 +44,7 @@ Possible configurations are:
 
 * Any time: Do not postpone deliveries
 * Fixed time windows: Postpone deliveries to the next preferred time window
+* Weekdays: Postpone deliveries to the next weekday
```
Comment (sebalix) [inline stock_partner_delivery_window/README.rst:47]: <s>Do not edit directly `README.rst`, it is automatically generated from files in `./readme/` subfolder. `CONFIGURE.rst` is the one you are looking for.</s>
nevermind, it's OK.

File: stock_partner_delivery_window/models/res_partner.py [hunk 5]
```diff
@@ -49,19 +57,13 @@ def get_delivery_windows(self, day_name=None):
...
             domain.append(("time_window_weekday_ids", "in", week_day_id))
         windows = self.env["partner.delivery.time.window"].search(domain)
-        for window in windows:
-            if not res.get(window.partner_id.id):
-                res[window.partner_id.id] = self.env[
-                    "partner.delivery.time.window"
-                ].browse()
-            res[window.partner_id.id] |= window
-        return res
+        return windows
```
Comment (sebalix) [inline stock_partner_delivery_window/models/res_partner.py]: Why did you changed this? The docstring says _"Return the list of delivery windows by partner id for the given day"_, maybe it has been done this way on purpose (overridden somewhere by custom code), so by changing the returned value you could break things.
Better to do such things while migrating the module to the next version.

File: stock_partner_delivery_window/models/res_partner.py [hunk 6]
```diff
@@ -71,68 +73,69 @@ def is_in_delivery_window(self, date_time):
...
+            else:
+                utc_start = start_time
+                utc_end = end_time
+            if utc_start <= date_time.time() < utc_end:
+                return True
         return False
 
     def _get_delivery_time_format_string(self):
-        return _("From %s to %s")
+        return _("From {} to {}")
```
Comment (sebalix) [inline stock_partner_delivery_window/models/res_partner.py]: /!\ same, this method could have been customized locally, better to not change this.

File: stock_partner_delivery_window/models/res_partner.py (hunk 1 above)
Comment (mmequignon) [reply stock_partner_delivery_window/models/res_partner.py]: Since we don't want to modify the behavior of existing methods, I reverted this change.

**Review by sebalix — APPROVED** [review]

**Comment by simahawk** [comment]
@i-vyshnevska can you update your review?
@mmequignon can you clean up commits?

**Review by i-vyshnevska — APPROVED** [review]

**Pull request merged**